include/oneapi/dpl/pstl/hetero/algorithm_ranges_impl_hetero.h

```diff
@@ -303,8 +414,30 @@ __pattern_adjacent_find(__hetero_tag<_BackendTag>, _ExecutionPolicy&& __exec, _R
 
     // inverted conditional because of
     // reorder_predicate in glue_algorithm_impl.h
-    return return_value(result, __rng.size(), __is__or_semantic);
+    if constexpr (__is__or_semantic())
+        return result ? 0 : __rng.size();
+    else
+        return result == __rng.size() - 1 ? __rng.size() : result;
+}
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+template <typename _BackendTag, typename _ExecutionPolicy, typename _R, typename _Proj, typename _Pred>
+auto
+__pattern_adjacent_find_ranges(__hetero_tag<_BackendTag> __tag, _ExecutionPolicy&& __exec, _R&& __r, _Pred __pred,
+                        _Proj __proj)
+{
+    auto __pred_2 = [__pred, __proj](auto&& __val, auto&& __next) 
+        { return std::invoke(__pred, std::invoke(__proj, std::forward<decltype(__val)>(__val)),
+        std::invoke(__proj, std::forward<decltype(__next)>(__next)));};
+
+    auto __idx =
+        oneapi::dpl::__internal::__ranges::__pattern_adjacent_find(__tag, std::forward<_ExecutionPolicy>(__exec),
+        oneapi::dpl::__ranges::views::all_read(std::forward<_R>(__r)), __pred_2,
+        oneapi::dpl::__internal::__first_semantic());
+
+    return std::ranges::borrowed_iterator_t<_R>(std::ranges::begin(__r) + __idx);
```

Comment:
The same thing with using already forwarded `__r`. that I've spotted. We need to carefully check the rest.

I already can see others.

Reply:
I was not find such cases. It seems they are already fixed.